File: common/src/main/java/cloudcmd/common/MetaUtil.java
package cloudcmd.common;


import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class MetaUtil
{
  public static JSONArray toJsonArray(List<FileMetaData> meta)
      throws JSONException
  {
    JSONArray result = new JSONArray();

    for (FileMetaData metaData : meta)
    {
      result.put(metaData.toJson());
    }

    return result;
  }

  // TODO: support file subblocks
  public static FileMetaData createMeta(File file, List<String> blockHashes, Set<String> tags)
      throws IOException, JSONException
  {

    String fileName = file.getName();

    int extIndex = fileName.lastIndexOf(".");

    FileMetaData meta = FileMetaData.create(
      JsonUtil.createJsonObject(
        "path", file.getCanonicalPath(),
        "filename", fileName,
        "fileext", extIndex >= 0 ? fileName.substring(extIndex + 1) : null,
        "filesize", file.length(),
        "filedate", file.lastModified(),
        "blocks", new JSONArray(blockHashes),
        "tags", new JSONArray(tags)
      ));

    return meta;
  }

  public static FileMetaData loadMeta(JSONObject jsonObject)
      throws JSONException, IOException
  {
    return loadMeta(jsonObject.getString("hash"), jsonObject.getJSONObject("data"));
  }

  public static FileMetaData loadMeta(String hash, JSONObject data)
      throws JSONException
  {
    return FileMetaData.create(hash, data);
  }

  public static FileMetaData deriveMeta(String hash, JSONObject data)
      throws JSONException, IOException
  {
    JSONObject derivedObj = new JSONObject(data.toString()); // TODO: there has to be a better way to clone!
    derivedObj.put("parent", hash);
    String derivedHash = CryptoUtil.computeHashAsString(new ByteArrayInputStream(derivedObj.toString().getBytes("UTF-8"))) + ".meta";
    FileMetaData meta = FileMetaData.create(derivedHash, derivedObj);
    return meta;
  }

  public static Set<String> prepareTags(List<String> incomingTags)
  {
    Set<String> tags = new HashSet<String>();

    for (String tag : incomingTags)
    {
      tag = tag.trim();
      if (tag.length() == 0) continue;
      String[] parts = tag.split(",");
      for (String part : parts)
      {
        part = part.trim();
        if (part.length() == 0) continue;
        tags.add(part);
      }
    }

    return tags;
  }

  public static Set<String> applyTags(Set<String> tags, Set<String> newTags)
  {
    Set<String> addTags = new HashSet<String>();
    Set<String> removeTags = new HashSet<String>();

    for (String tag : newTags)
    {
      if (tag.startsWith("-"))
      {
        removeTags.add(tag.substring(1));
      }
      else
      {
        addTags.add(tag);
      }
    }

    Set<String> appliedTags = new HashSet<String>(tags);

    appliedTags.addAll(addTags);
    appliedTags.removeAll(removeTags);

    return appliedTags;
  }
}

File: common/src/main/scala/cloudcmd/common/util/MetaUtil.scala
package cloudcmd.common.util

import java.io.ByteArrayInputStream
import java.io.File
import org.json.JSONArray
import org.json.JSONObject

object MetaUtil {
  def toJsonArray(meta: List[FileMetaData]): JSONArray = {
    val result = new JSONArray
    meta.foreach(m => result.put(m.toJson))
    result
  }

  def createMeta(file: File, blockHashes: List[String], tags: Set[String]): FileMetaData = {
    val fileName = file.getName
    val extIndex = fileName.lastIndexOf(".")
    FileMetaData.create(
      JsonUtil.createJsonObject(
        "path", file.getCanonicalPath,
        "filename", fileName,
        "fileext", if (extIndex >= 0) fileName.substring(extIndex + 1) else null,
        "filesize", file.length.asInstanceOf[AnyRef],
        "filedate", file.lastModified.asInstanceOf[AnyRef],
        "blocks", JsonUtil.toJsonArray(blockHashes),
        "tags", JsonUtil.toJsonArray(tags)))
  }

  def loadMeta(jsonObject: JSONObject): FileMetaData = {
    loadMeta(jsonObject.getString("hash"), jsonObject.getJSONObject("data"))
  }

  def loadMeta(hash: String, data: JSONObject): FileMetaData = {
    FileMetaData.create(hash, data)
  }

  def deriveMeta(hash: String, data: JSONObject): FileMetaData = {
    val derivedObj: JSONObject = new JSONObject(data.toString)
    derivedObj.put("parent", hash)
    val derivedHash: String = CryptoUtil.computeHashAsString(new ByteArrayInputStream(derivedObj.toString.getBytes("UTF-8"))) + ".meta"
    val meta: FileMetaData = FileMetaData.create(derivedHash, derivedObj)
    meta
  }

  def prepareTags(incomingTags: List[String]): Set[String] = {
    incomingTags.map(_.trim).filter(_.length > 0).flatMap(_.split(",").filter(_.length > 0).toSet).toSet
  }

  def applyTags(tags: Set[String], newTags: Set[String]): Set[String] = {
    val modTags = prepareTags(newTags.toList)
    val subTags = modTags.filter(_.startsWith("-"))
    val rmTags = subTags.map(_.substring(1)).toSet
    val addTags = modTags -- subTags
    (tags -- rmTags) ++ addTags
  }
}